File: cpp/vm/advm/ast.cpp
#include "ast.h"

std::string ASTNode::toString() {
    return "unimplemented in subclass";
}

std::string ASTNode::hash() {
    return "unimplemented in subclass";
}

std::string ASTNode::inspect() {
    return "unimplemented in subclass";
}

ASTProgram::ASTProgram() {
    type = AT_PROGRAM;
    ref_count = 0;
}

ASTProgram::~ASTProgram() {
    for(std::vector<ASTNode*>::iterator it = statements.begin() ; it != statements.end(); ++it) {
        ASTNode *node = *it;
        free_memory_ASTNode(node);
    }
}

void ASTProgram::reset() {
    for(std::vector<ASTNode*>::iterator it = statements.begin() ; it != statements.end(); ++it) {
        ASTNode *node = *it;
        free_memory_ASTNode(node);
    }
    statements.clear();
}

std::string ASTProgram::toString() {
    std::string out = "";
    out += "AstProgram\n";
    for (ASTNode* it : statements) {
        out += it->toString();
    }
    return out;
}

std::string ASTProgram::inspect() {
    return "AstProgram\n";
}

ASTExpressionStatement::ASTExpressionStatement() {
    type = AT_EXPRESSION_STATEMENT;
    ref_count = 0;
}

ASTExpressionStatement::ASTExpressionStatement(Token t) {
    type = AT_EXPRESSION_STATEMENT;
    ref_count = 0;
    token = t;
}

ASTExpressionStatement::ASTExpressionStatement(Token t, ASTNode* e) {
    type = AT_EXPRESSION_STATEMENT;
    ref_count = 0;
    token = t;
    expression = e;
}

ASTExpressionStatement::~ASTExpressionStatement() {
    free_memory_ASTNode(expression);
}

std::string ASTExpressionStatement::toString() {
    if (expression == NULL) {
        return "ASTExpressionStatement[None]\n";
    }
    return "ASTExpressionStatement[" + expression->toString() + "]\n";
}

std::string ASTExpressionStatement::inspect() {
    return "ASTExpressionStatement";
}

ASTInfixExpression::ASTInfixExpression() {
    type = AT_INFIX_EXPRESSION;
    ref_count = 0;
}

ASTInfixExpression::ASTInfixExpression(Token t) {
    type = AT_INFIX_EXPRESSION;
    ref_count = 0;
    token = t;
}

ASTInfixExpression::ASTInfixExpression(Token t, ASTNode* l, ASTNode* r, std::string op) {
    type = AT_INFIX_EXPRESSION;
    ref_count = 0;
    token = t;
    left = l;
    right = r;
    _operator = op;
}

ASTInfixExpression::~ASTInfixExpression() {
    free_memory_ASTNode(left);
    free_memory_ASTNode(right);
}

std::string ASTInfixExpression::toString() {
    return "ASTInfixExpression";
}

std::string ASTInfixExpression::inspect() {
    return "ASTInfixExpression";
}

ASTIdentifier::ASTIdentifier() {
    type = AT_IDENTIFIER;
    ref_count = 0;
}

ASTIdentifier::ASTIdentifier(Token t, std::string val) {
    type = AT_IDENTIFIER;
    ref_count = 0;
    token = t;
    value = val;
}

std::string ASTIdentifier::toString() {
    return "ASTIdentifier[" + token.stringLiteral + "]";
}

std::string ASTIdentifier::inspect() {
    return "ASTIdentifier";
}

ASTWhileStatement::ASTWhileStatement() {

}

ASTWhileStatement::ASTWhileStatement(Token t) {

}

std::string ASTWhileStatement::inspect() {
    return "todo: implement this";
}

std::string ASTWhileStatement::toString() {
    return "todo: implement this";
}

ASTInteger::ASTInteger() {
    type = AT_INTEGER;
    ref_count = 0;
}

ASTInteger::ASTInteger(Token t) {
    type = AT_INTEGER;
    ref_count = 0;
    token = t;
}

ASTInteger::ASTInteger(Token t, int v) {
    type = AT_INTEGER;
    ref_count = 0;
    token = t;
    value = v;
}

ASTInteger::~ASTInteger() {
    //...
}

std::string ASTInteger::toString() {
    return "ASTInteger[" + std::to_string(value) + "]";
}

std::string ASTInteger::inspect() {
    return "ASTInteger";
}

/*std::string ASTInteger::hash() {
    return "test";
}*/

ASTPrefixExpression::ASTPrefixExpression(Token t, std::string o) {
    type = AT_PREFIX_EXPRESSION;
    ref_count = 0;
    token = t;
    _operator = o;
}

ASTPrefixExpression::~ASTPrefixExpression() {
    free_memory_ASTNode(right);
}

ASTBoolean::ASTBoolean() {
    type = AT_BOOLEAN;
    ref_count = 0;
}

ASTBoolean::ASTBoolean(Token t) {
    type = AT_BOOLEAN;
    ref_count = 0;
    token = t;
}

ASTBoolean::ASTBoolean(Token t, bool v) {
    type = AT_BOOLEAN;
    ref_count = 0;
    token = t;
    value = v;
}

ASTBoolean::~ASTBoolean() {
    //...
}

std::string ASTBoolean::toString() {
    return "ASTBoolean";
}

std::string ASTBoolean::inspect() {
    return "ASTBoolean";
}

ASTIfStatement::ASTIfStatement() {
    type = AT_IF_STATEMENT;
    ref_count = 0;
    condition = nullptr;
    consequence = nullptr;
    alternative = nullptr;
}

ASTIfStatement::ASTIfStatement(Token t) {
    type = AT_IF_STATEMENT;
    ref_count = 0;
    token = t;
    condition = nullptr;
    consequence = nullptr;
    alternative = nullptr;
}

ASTIfStatement::~ASTIfStatement() {
    free_memory_ASTNode(condition);
    free_memory_ASTNode(consequence);
    free_memory_ASTNode(alternative);
}

std::string ASTIfStatement::inspect() {
    return "todo: implement this";
}

std::string ASTIfStatement::toString() {
    return "ASTIfExpression";
}

ASTBlockStatement::ASTBlockStatement() {
    type = AT_BLOCK_STATEMENT;
    ref_count = 0;
}

ASTBlockStatement::ASTBlockStatement(Token t) {
    type = AT_BLOCK_STATEMENT;
    ref_count = 0;
    token = t;
}

ASTBlockStatement::~ASTBlockStatement() {
    for (auto& it : statements) {
        free_memory_ASTNode(it);
    }
}

std::string ASTBlockStatement::inspect() {
    std::string out = "{";
    for (auto& it : statements) {
        out += it->inspect();
    }
    out += "}";
    return out;
}

std::string ASTBlockStatement::toString() {
    std::string out = "{";
    for (auto& it : statements) {
        out += it->inspect();
    }
    out += "}";
    return out;
}

ASTNullExpression::ASTNullExpression() {
    type = AT_NULL_EXPRESSION;
    ref_count = 0;
}

std::string ASTNullExpression::inspect() {
    return "ASTNullExpression";
}

std::string ASTNullExpression::toString() {
    return "ASTNullExpression";
}

ASTLetStatement::ASTLetStatement() {
    type = AT_LET_STATEMENT;
    ref_count = 0;
}

ASTLetStatement::ASTLetStatement(Token t) {
    type = AT_LET_STATEMENT;
    ref_count = 0;
    token = t;
}

ASTLetStatement::~ASTLetStatement() {
    if (value) {
        free_memory_ASTNode(value);
    }
}

std::string ASTLetStatement::inspect() {
    return "ASTLetStatement";
}

std::string ASTLetStatement::toString() {
    return "ASTLetStatement<" + name.toString() + "> [" + value->toString() + "]\n";
}

ASTReturnStatement::ASTReturnStatement() {
    type = AT_RETURN_STATEMENT;
    ref_count = 0;
}

ASTReturnStatement::ASTReturnStatement(Token t) {
    type = AT_RETURN_STATEMENT;
    ref_count = 0;
    token = t;
}

ASTReturnStatement::~ASTReturnStatement() {
    //...
}

std::string ASTReturnStatement::inspect() {
    return "todo: implement ASTReturnStatement.inspect()";
}

std::string ASTReturnStatement::toString() {
    return "todo: implement ASTReturnStatement.toString()";
}

ASTString::ASTString() {
    type = AT_STRING_LITERAL;
    ref_count = 0;
}

ASTString::ASTString(Token t) {
    type = AT_STRING_LITERAL;
    ref_count = 0;
    token = t;
}

ASTString::ASTString(Token t, std::string v) {
    type = AT_STRING_LITERAL;
    ref_count = 0;
    token = t;
    value = v;
}

ASTString::~ASTString() {
    //...
}

std::string ASTString::inspect() {
    return "ASTString.inspect()";
}

std::string ASTString::toString() {
    return "ASTString[" + value + "]";
}

ASTList::ASTList() {
    type = AT_LIST_LITERAL;
    ref_count = 0;
}

ASTList::ASTList(Token t) {
    type = AT_LIST_LITERAL;
    ref_count = 0;
    token = t;
}

ASTList::ASTList(Token t, std::vector<ASTNode*> e) {
    type = AT_LIST_LITERAL;
    ref_count = 0;
    token = t;
    elements = e;
}

ASTList::~ASTList() {
    for (auto& it : elements) {
        free_memory_ASTNode(it);
    }
}

std::string ASTList::inspect() {
    return "ASTList.inspect()";
}

std::string ASTList::toString() {
    std::string out = "ASTList[";
    int i = 0, size = elements.size();
    for (auto el : elements) {
        if (i < size - 1) {
            out += el->inspect() + ", ";
        } else {
            out += el->inspect();
        }
        ++i;
    }
    out += "]";
    return out;
}

ASTIndexExpression::ASTIndexExpression() {
    type = AT_INDEX_EXPRESSION;
    ref_count = 0;
}

ASTIndexExpression::ASTIndexExpression(Token t) {
    type = AT_INDEX_EXPRESSION;
    ref_count = 0;
    token = t;
}

ASTIndexExpression::ASTIndexExpression(Token t, ASTNode *l) {
    type = AT_INDEX_EXPRESSION;
    ref_count = 0;
    token = t;
    left = l;
}

ASTIndexExpression::~ASTIndexExpression() {
    //...
}

std::string ASTIndexExpression::inspect() {
    return "ASTIndexExpression.inspect()";
}

std::string ASTIndexExpression::toString() {
    std::string out = "ASTIndexExpression[";
    out += "<" + left->toString() + ">";
    out += "<" + index->toString() + ">";
    if (indexEnd) {
        out += "<" + indexEnd->toString() + ">";
    }
    if (step) {
        out += "<" + step->toString() + ">";
    }
    out += "]";
    return out;
}

ASTHash::ASTHash() {
    type = AT_HASH_LITERAL;
    ref_count = 0;
}

ASTHash::ASTHash(Token t) {
    type = AT_HASH_LITERAL;
    ref_count = 0;
    token = t;
}

ASTHash::ASTHash(Token t, std::unordered_map<ASTNode*, ASTNode*> p) {
    type = AT_HASH_LITERAL;
    ref_count = 0;
    token = t;
    pairs = p;
}

ASTHash::~ASTHash() {
    for (auto& it : pairs) {
        free_memory_ASTNode(it.first);
        free_memory_ASTNode(it.second);
    }
}

std::string ASTHash::inspect() {
    return "TODO: implement ASTHash.inspect()";
}

std::string ASTHash::toString() {
    std::string out = "ASTHashLiteral[";
    for (auto& it: pairs) {
        out += "<" + it.first->toString() + ": " + it.second->toString() + ">";
    }
    out += "]";
    return out;
}

ASTFunctionLiteral::ASTFunctionLiteral() {
    type = AT_FUNCTION_LITERAL;
    ref_count = 0; // poate aici ar trebui sa fie 1
}

ASTFunctionLiteral::ASTFunctionLiteral(Token t) {
    type = AT_FUNCTION_LITERAL;
    ref_count = 0; // poate aici ar trebui sa fie 1
    token = t;
}

ASTFunctionLiteral::~ASTFunctionLiteral() {
    if (body) {
        Ad_DECREF(body); // asta merge si e super cool
        free_memory_ASTNode(body);
    }
    for (std::vector<ASTNode*>::iterator it = parameters.begin() ; it != parameters.end(); ++it) {
        ASTNode *node = *it;
        Ad_DECREF(node); // asta merge si e super cool
        free_memory_ASTNode(node);
    }
    for (std::vector<ASTNode*>::iterator it = default_params.begin() ; it != default_params.end(); ++it) {
        ASTNode *node = *it;
        Ad_DECREF(node); // asta merge si e super cool
        free_memory_ASTNode(node);
    }
}

std::string ASTFunctionLiteral::inspect() {
    return "TODO: implement ASTFunctionLiteral.inspect()";
}

std::string ASTFunctionLiteral::toString() {
    std::string out = "[FunctionLiteral] <";
    for (std::vector<ASTNode*>::iterator it = parameters.begin() ; it != parameters.end(); ++it) {
        ASTNode *current = *it;
        if (it != parameters.begin()) out += ", ";
        out += current->toString();
    }
    out += "><";
    out += body->toString();
    out += ">";
    return out;
}

ASTCallExpression::ASTCallExpression() {
    type = AT_CALL_EXPRESSION;
}

ASTCallExpression::ASTCallExpression(Token t) {
    type = AT_CALL_EXPRESSION;
    token = t;
}

ASTCallExpression::ASTCallExpression(Token t, ASTNode* f) {
    type = AT_CALL_EXPRESSION;
    token = t;
    function = f;
}

ASTCallExpression::~ASTCallExpression() {
    free_memory_ASTNode(function);
    for (std::vector<ASTNode*>::iterator it = arguments.begin() ; it != arguments.end(); ++it) {
        ASTNode *node = *it;
        free_memory_ASTNode(node);
    }
    for (std::vector<ASTNode*>::iterator it = kw_args.begin() ; it != kw_args.end(); ++it) {
        ASTNode *node = *it;
        free_memory_ASTNode(node);
    }
}

std::string ASTCallExpression::inspect() {
    return "todo: implement ASTCallExpression.inspect()";
}

std::string ASTCallExpression::toString() {
    return "todo: implement ASTCallExpression.toString()";
}

ASTNode* ASTCallExpression::copy() {
    return nullptr;
}

void Ad_INCREF(ASTNode* node) {
    if (node) {
        node->ref_count++;
    }
}

void Ad_DECREF(ASTNode* node) {
    if (node) {
        node->ref_count--;
    }
}

void free_memory_ASTNode(ASTNode* node) {
    if (node == NULL) return;
    if (node->ref_count > 0) return;
    switch(node->type)  {
        case AT_PROGRAM: {
            delete (ASTProgram*) node;
            break;
        }
        case AT_EXPRESSION_STATEMENT: {
            delete (ASTExpressionStatement*) node;
            break;
        }
        case AT_IDENTIFIER: {
            delete (ASTIdentifier*) node;
            break;
        }
        case AT_INTEGER: {
            delete (ASTInteger*) node;
            break;
        }
        case AT_BOOLEAN: {
            delete (ASTBoolean*) node;
            break;
        }
        case AT_INFIX_EXPRESSION: {
            delete (ASTInfixExpression*) node;
            break;
        }
        case AT_PREFIX_EXPRESSION: {
            delete (ASTPrefixExpression*) node;
            break;
        }
        case AT_LET_STATEMENT: {
            delete (ASTLetStatement*) node;
            break;
        }
        case AT_RETURN_STATEMENT: {
            delete (ASTReturnStatement*) node;
            break;
        }
        case AT_FUNCTION_STATEMENT: {
            delete (ASTFunctionStatement*) node;
            break;
        }
        case AT_DEF_STATEMENT: {
            delete (ASTDefStatement*) node;
            break;
        }
        case AT_WHILE_STATEMENT: {
            delete (ASTWhileStatement*) node;
            break;
        }
        case AT_FOR_STATEMENT: {
            delete (ASTForStatement*) node;
            break;
        }
        case AT_IF_STATEMENT: {
            delete (ASTIfStatement*) node;
            break;
        }
        case AT_BLOCK_STATEMENT: {
            delete (ASTBlockStatement*) node;
            break;
        }
        case AT_NULL_EXPRESSION: {
            delete (ASTNullExpression*) node;
            break;
        }
        case AT_STRING_LITERAL: {
            delete (ASTString*) node;
            break;
        }
        case AT_LIST_LITERAL: {
            delete (ASTList*) node;
            break;
        }
        case AT_HASH_LITERAL: {
            delete (ASTHash*) node;
            break;
        }
        case AT_INDEX_EXPRESSION: {
            delete (ASTIndexExpression*) node;
            break;
        }
        case AT_FUNCTION_LITERAL: {
            delete (ASTFunctionLiteral*) node;
            break;
        }
        case AT_CALL_EXPRESSION: {
            delete (ASTCallExpression*) node;
            break;
        }
        default: {
            std::cout << "SEVERE MEMORY ERROR: " << ast_type_map[node->type] << "!!!\n";
            break;
        }
    }
}

File: python/VMPrototype1/ast.py
from typing import List, Dict

from token_ad import Token


class StatementType:
    PROGRAM = 'PROGRAM'
    LET_STATEMENT = 'LET_STATEMENT'
    RETURN_STATEMENT = 'RETURN_STATEMENT'
    EXPRESSION_STATEMENT = 'EXPRESSION_STATEMENT'
    IDENTIFIER = 'IDENTIFIER'
    INTEGER_LITERAL = 'INTEGER_LITERAL'
    FLOAT = 'FLOAT'
    BOOLEAN = 'BOOLEAN'
    INFIX_EXPRESSION = 'INFIX_EXPRESSION'
    PREFIX_EXPRESSION = 'PREFIX_EXPRESSION'
    CALL_EXPRESSION = 'CALL_EXPRESSION'
    IF_EXPRESSION = 'IF_EXPRESSION'
    BLOCK_STATEMENT = 'BLOCK_STATEMENT'
    FUNCTION_LITERAL = 'FUNCTION_LITERAL'
    WHILE_EXPRESSION = 'WHILE_EXPRESSION'
    STRING_LITERAL = 'STRING_LITERAL'
    LIST_LITERAL = 'LIST_LITERAL'
    INDEX_EXPRESSION = 'INDEX_EXPRESSION'
    HASH_LITERAL = 'HASH_LITERAL'
    ASSIGN_STATEMENT = 'ASSIGN_STATEMENT'
    DEF_STATEMENT = 'DEF_STATEMENT'
    CLASS_STATEMENT = 'CLASS_STATEMENT'
    MEMBER_ACCESS = 'MEMBER_ACCESS'
    COMMENT = 'COMMENT'
    PREFIX_INCREMENT = 'PREFIX_INCREMENT'
    POSTFIX_INCREMENT = 'POSTFIX_INCREMENT'
    FOR_EXPRESSION = 'FOR_EXPRESSION'
    NULL_EXPRESSION = 'NULL_EXPRESSION'
    THIS_EXPRESSION = 'THIS_EXPRESSION'
    SUPER_EXPRESSION = 'SUPER_EXPRESSION'
    BREAK_STATEMENT = 'BREAK_STATEMENT'
    CONTINUE_STATEMENT = 'CONTINUE_STATEMENT'
    PLUS_EQUALS_STATEMENT = 'ASSIGN_EQUALS_STATEMENT'


statement_type_map = {
    StatementType.PROGRAM: 'PROGRAM',
    StatementType.EXPRESSION_STATEMENT: 'EXPRESSION_STATEMENT',
    StatementType.INTEGER_LITERAL: 'INTEGER_LITERAL',
    StatementType.INFIX_EXPRESSION: 'INFIX_EXPRESSION',
    StatementType.PREFIX_EXPRESSION: 'PREFIX_EXPRESSION',
    StatementType.BOOLEAN: 'BOOLEAN',
    StatementType.IF_EXPRESSION: 'IF_EXPRESSION',
    StatementType.BLOCK_STATEMENT: 'BLOCK_STATEMENT',
    StatementType.NULL_EXPRESSION: 'NULL_EXPRESSION',
    StatementType.LET_STATEMENT: 'LET_STATEMENT',
    StatementType.IDENTIFIER: 'IDENTIFIER',
    StatementType.STRING_LITERAL: 'STRING_LITERAL',
    StatementType.LIST_LITERAL: 'LIST_LITERAL',
    StatementType.INDEX_EXPRESSION: 'INDEX_EXPRESSION',
    StatementType.HASH_LITERAL: 'HASH_LITERAL',
    StatementType.FUNCTION_LITERAL: 'FUNCTION_LITERAL',
    StatementType.CALL_EXPRESSION: 'CALL_EXPRESSION',
    StatementType.ASSIGN_STATEMENT: 'ASSIGN_STATEMENT',
    StatementType.RETURN_STATEMENT: 'RETURN_STATEMENT',
    StatementType.WHILE_EXPRESSION: 'WHILE_EXPRESSION'
}


class ASTNode:
    statement_type = None

    def __init__(self):
        pass

    def token_literal(self):
        print('token_literal unimplemented in subclass')

    def __str__(self):
        print('__str__ unimplemented in subclass')


class ASTProgram(ASTNode):
    statement_type = StatementType.PROGRAM

    def __init__(self):
        self.statements = []

    def reset(self):
        self.statements = []

    def token_literal(self):
        return "PROGRAM"

    def __str__(self):
        return 'ASTProgram\n' + '\n'.join([str(stmt) for stmt in self.statements])


class ASTExpressionStatement(ASTNode):
    statement_type = StatementType.EXPRESSION_STATEMENT

    def __init__(self, token: Token, expression: ASTNode = None):
        self.token = token
        self.expression = expression

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTExpressionStatement[' + str(self.expression) + ']'


class ASTInteger(ASTNode):
    statement_type = StatementType.INTEGER_LITERAL

    def __init__(self, token: Token, value: int = None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTInteger[' + str(self.value) + ']'


class ASTString(ASTNode):
    statement_type = StatementType.STRING_LITERAL

    def __init__(self, token: Token, value: str = None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTString[' + self.value + ']'


class ASTList(ASTNode):
    statement_type = StatementType.LIST_LITERAL

    def __init__(self, token: Token = None, elements: List = None):
        super().__init__()
        self.token = token
        self.elements = elements

    def token_literal(self) -> str:
        return self.token.literal

    def __str__(self) -> str:
        return '[' + ','.join([str(element) for element in self.elements]) + ']'


class ASTIndexExpression(ASTNode):
    statement_type = StatementType.INDEX_EXPRESSION

    def __init__(self, token: Token = None, left: ASTNode = None, index: ASTNode = None):
        super().__init__()
        self.token = token
        self.left = left
        self.index = index

    def token_literal(self) -> str:
        return self.token.literal

    def __str__(self) -> str:
        return 'ASTIndexExpression<' + str(self.left) + '>[' + str(self.index) + ']'


class ASTHash(ASTNode):
    statement_type = StatementType.HASH_LITERAL

    def __init__(self, token: Token = None, pairs: Dict[ASTNode, ASTNode] = None):
        """
        @param token: the node's token
        @param pairs: map [Expression] => Expression
        """
        super().__init__()
        self.token = token
        self.pairs = pairs

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        out = 'ASTHashLiteral['
        for key, value in self.pairs.items():
            out += "<" + str(key) + ": " + str(value) + ">"
        out += "]"
        return out


class ASTInfixExpression(ASTNode):
    statement_type = StatementType.INFIX_EXPRESSION

    def __init__(self, token=None, operator=None, left=None, right=None):
        self.token = token
        self.operator = operator
        self.left = left
        self.right = right

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return "ASTInfixExpression [{0} <{1}> {2}]".format(str(self.left), str(self.operator), str(self.right))


class ASTPrefixExpression(ASTNode):
    statement_type = StatementType.PREFIX_EXPRESSION

    def __init__(self, token=None, operator=None, right=None):
        self.token = token
        self.operator = operator
        self.right = right

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return "ASTPrefixExpression <{0}> [{1}]".format(self.operator, self.right)


class ASTBoolean(ASTNode):
    statement_type = StatementType.BOOLEAN

    def __init__(self, token: Token, value: bool = None):
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTBoolean[' + str(self.value) + ']'


class ASTIfExpression(ASTNode):
    statement_type = StatementType.IF_EXPRESSION

    def __init__(self, token=None, condition=None, consequence=None, alternative=None):
        """
        @param token: the node's token
        @param condition: expression???
        @param consequence: ASTBlockStatement
        @param alternative: ASTBlockStatement
        """
        self.token = token
        self.condition = condition
        self.consequence = consequence
        self.alternative = alternative

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTIfExpression'


class ASTBlockStatement(ASTNode):
    statement_type = StatementType.BLOCK_STATEMENT

    def __init__(self, token=None, statements=None):
        """
        @param token: the node's token
        @param statements: list
        """
        self.token = token
        if statements:
            self.statements = statements
        else:
            self.statements = []

    def __str__(self):
        out = ""
        for statement in self.statements:
            out += str(statement)
        return out


class ASTLetStatement(ASTNode):
    statement_type = StatementType.LET_STATEMENT

    def __init__(self, token=None, name=None, value=None):
        self.token = token
        self.name = name
        self.value = value

    def __str__(self):
        return 'ASTLetStatement<' + str(self.name) + '> [' + str(self.value) + ']'


class ASTFunctionLiteral(ASTNode):
    statement_type = StatementType.FUNCTION_LITERAL

    def __init__(self, token=None, parameters=None, default_params=None, body=None, name: str = None):
        super().__init__()
        self.token = token
        self.parameters = parameters
        self.default_params = default_params
        self.body = body
        self.name = name

    def __str__(self):
        return 'ASTFunctionLiteral'


class ASTCallExpression(ASTNode):
    statement_type = StatementType.CALL_EXPRESSION

    def __init__(self, token=None, func=None):
        super().__init__()
        self.token = token
        self.func = func
        self.arguments = None
        self.kw_args = None

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTCallExpression[some arguments here]'


class ASTReturnStatement(ASTNode):
    statement_type = StatementType.RETURN_STATEMENT

    def __init__(self, token: Token = None, value: ASTNode = None):
        super().__init__()
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTReturnStatement[some arguments here]'


class ASTIdentifier(ASTNode):
    statement_type = StatementType.IDENTIFIER

    def __init__(self, token: Token = None, value: str = None):
        super().__init__()
        self.token = token
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'ASTIdentifier[' + self.value + ']'


class ASTNullExpression(ASTNode):
    statement_type = StatementType.NULL_EXPRESSION

    def __init__(self, token=None):
        self.token = token

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return "ASTNullExpression"


class ASTThisExpression(ASTNode):
    statement_type = StatementType.THIS_EXPRESSION

    def __init__(self, token=None):
        """
        @param token: the node's token
        """
        super().__init__()
        self.token = token

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'Token: ' + str(self.token)


class ASTMemberAccess(ASTNode):
    statement_type = StatementType.MEMBER_ACCESS

    def __init__(self, token=None, arguments=None, kw_args=None, owner=None, member=None, is_method=False):
        """
        @param token: the node's token
        @param arguments:
        @param owner:
        @param member:
        @param is_method: boolean, indicates is this member access is a method call or an attribute access
        """
        super().__init__()
        self.token = token
        self.arguments = arguments
        self.kw_args = kw_args
        self.owner = owner
        self.member = member
        self.is_method = is_method

    def __str__(self):
        return "ASTMemberAccess{owner: " + str(self.owner) + ", member: " + str(self.member) + "}"


class ASTDefStatement(ASTNode):
    statement_type = StatementType.DEF_STATEMENT

    def __init__(self, token: Token = None, name=None, parameters=None, default_params=None, body=None):
        """
        @param name: AstNode - identifier that will store the function, it needs to be like a let statement in the eval
        @param parameters: AstNode[] list of funtion argumentss
        @:param default_params: default params when missing argument
        @param body: AstNode block statement that is the body of the function
        """
        super().__init__()
        self.name = name
        self.token = token
        self.parameters = parameters
        self.default_params = default_params
        self.body = body

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'Token: ' + str(self.token) + '<name: ' + str(self.name) + ', params:' + str(
            self.parameters) + ', body: ' + str(self.body) + '>'


class ASTAssignStatement(ASTNode):
    statement_type = StatementType.ASSIGN_STATEMENT

    def __init__(self, token: Token = None, name: ASTNode = None, value: ASTNode = None):
        super().__init__()
        self.token = token
        self.name = name
        self.value = value

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        if self.name.statement_type == StatementType.MEMBER_ACCESS:
            return 'AssignStatement [' + str(self.token.literal) + '] <' + str(self.name) + '>: ' + (
                str(self.value) if self.value else '')
        return 'AssignStatement [' + str(self.token.literal) + '] <' + str(self.name.value) +'>: ' + (str(self.value) if self.value else '')


class ASTClassStatement(ASTNode):
    statement_type = StatementType.CLASS_STATEMENT

    def __init__(self, token: Token = None, name: ASTNode = None, methods: List[ASTDefStatement] = None,
                 attributes: List[ASTAssignStatement] = None, inherit_from: List[ASTNode] = None):
        """
        @param token: the node's token
        @param name: ASTIdentifier, name of the class
        @param methods: list of ASTDefStatement
        @param attributes: list of ASTAssignStatement
        @param inherit_from: list of parents the class inherits from, should this be a list of ASTNode(s)?
        """
        super().__init__()
        self.token = token
        self.name = name
        self.methods = methods
        self.attributes = attributes
        self.inherit_from = inherit_from

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'Token: ' + str(self.token)


class ASTWhileExpression(ASTNode):
    statement_type = StatementType.WHILE_EXPRESSION

    def __init__(self, token: Token = None, condition: ASTNode = None, block: ASTNode = None):
        """
        @param token: the node's token
        @param condition: expression???
        @param block: ASTBlockStatement
        """
        super().__init__()
        self.token = token
        self.condition = condition
        self.block = block

    def token_literal(self):
        return self.token.literal

    def __str__(self):
        return 'TODO: implement __str__ in ASTWhileExpression'
